### PR TITLE
AA8K-17

### PR DESCRIPTION
**Pull request by hendrysiak**
First implementation of google auth.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import express from 'express';
+import { mongoDBConnectionString, mongoOptions, port } from './config';
+import { connect } from './connectToMongo';
```
Comment (hayuna) [inline src/app.ts]: Could we move everything related to mongo to module DBService ?

File: src/connectToMongo.ts [hunk 2]
```diff
@@ -13,6 +13,7 @@ const defaultOptions: IMongoOptions = {
 async function connect(connectionString: string, options: IMongoOptions = defaultOptions): Promise<boolean> {
   try{
     await mongoose.connect(connectionString, options)
+    mongoose.set('useFindAndModify', false);
```
Comment (hayuna) [inline src/connectToMongo.ts]: We can set this value in options in DB Service

File: src/controllers/user.controller.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { Request, Response } from 'express';
+import { IUserAttached } from '../models/user/user.model';
+import { UserService } from '../services/user.service';
+
+const userService = new UserService();
```
Comment (hayuna) [inline src/controllers/user.controller.ts]: As we talked before. We can refactor this into exported object from module

File: src/controllers/user.controller.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+const getUserByEmail = (email: string): Promise<IUserAttached | null> => {
+  return userService.getUserData(email);
+}
+
+export const loginUserViaGoogle = async (verifiedUser: any, req: Request, res: Response): Promise<void> => {
```
Comment (hayuna) [inline src/controllers/user.controller.ts]: Could we change this `any` type into something different?

File: src/controllers/user.controller.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  if (!verifiedUser) res.status(401).send('User does not exist!');
+
+  console.log(verifiedUser);
+
+  const user = await getUserByEmail(verifiedUser.emails[0].value);
```
Comment (hayuna) [inline src/controllers/user.controller.ts:16]: User can have more than one email?

File: src/controllers/user.controller.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  if (user === null) {
```
Comment (hayuna) [inline src/controllers/user.controller.ts]:
```suggestion
  if (!user) {
```

File: src/controllers/user.controller.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    const savedUser = await userService.saveUserData({userEmail: verifiedUser.emails[0].value, lastLoggedIn: new Date()});
```
Comment (hayuna) [inline src/controllers/user.controller.ts]: Could we use built-in timestamp methods in mongo?

File: src/models/user/user.model.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { IBaseModelAttached } from '../base-model';
+
+export interface IUser {
+
+  /** A user's email */
+  userEmail: string;
```
Comment (hayuna) [inline src/models/user/user.model.ts]: I think `email` would be enough

File: src/models/user/user.schema.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+export const UserCollectionName = 'Users';
+
+const UserSchema = new Schema({
+  userEmail: {
+    type: String,
+    required: true
+  },
+  lastLoggedIn: {
+      type: Date,
+      required: false
```
Comment (hayuna) [inline src/models/user/user.schema.ts]: By default required is false. Not need to specify it

File: src/models/user/user.schema.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+mongoose.model<IUserDocument>(UserCollectionName, UserSchema);
+
+export const mapUserToAttachedUser = (user: IUserDocument | null): IUserAttached | null => {
+  if (user) {
+    return {
+      id: user.id,
+      userEmail: user.userEmail,
+      lastLoggedIn: user.lastLoggedIn
+    };
+  } else return null;
```
Comment (hayuna) [inline src/models/user/user.schema.ts:29]:
```suggestion
  } 
  return null;
```

**Review by hayuna — CHANGES_REQUESTED** [review]
Comments above

File: src/controllers/user.controller.ts (hunk 5 above)
Comment (hendrysiak) [reply src/controllers/user.controller.ts:16]: Yes, and we should use main email - on first position.

File: src/controllers/user.controller.ts (hunk 7 above)
Comment (hendrysiak) [reply src/controllers/user.controller.ts]: This is just a tradeoff - we can, but we should change IUserAttached type to can obtain object with $currentDate mongoDB operator and type checking - few advantages, many disadvantages.

File: src/config/index.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+export const port = 3000;
+export const mongoDBConnectionString = process.env.MONGODBURL as string;
+export const mongoOptions = {
```
Comment (hayuna) [inline src/config/index.ts]: Do we still need this if we have db service?

File: src/services/db.service.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import { mongoDBConnectionString, mongoOptions } from '../config';
+
+import mongoose from 'mongoose'
+import { defaultOptions, IMongoOptions } from '../models/mongo/mongo.model';
+
+const connect = async(connectionString: string, options: IMongoOptions = defaultOptions): Promise<boolean> => {
+  try{
+    await mongoose.connect(connectionString, options)
+    mongoose.set('useFindAndModify', false);
```
Comment (hayuna) [inline src/services/db.service.ts]: Could we move it to defaultOptions?

**Pull request merged**